theano/gof/cmodule.py

```diff
@@ -1392,86 +1396,101 @@ def clear_unversioned(self, min_age=None):
-                        # crashed, and we take care of the clean-up.
-                        if age > min_age:
-                            _rmtree(os.path.join(self.dirname, filename),
-                                    msg='old unversioned', level=logging.INFO,
-                                    ignore_nocleanup=True)
+        to_del = []
+        time_now = time.time()
+        for filename in os.listdir(self.dirname):
+            if filename.startswith('tmp'):
+                try:
+                    fname = os.path.join(self.dirname, filename, 'key.pkl')
+                    open(fname).close()
+                    has_key = True
+                except IOError:
+                    has_key = False
+                if not has_key:
+                    # Use the compiled file by default
+                    path = module_name_from_dir(os.path.join(self.dirname,
+                                                             filename),
+                                                False)
+                    # If it don't exist, use any file in the directory.
+                    if path is None:
+                        path = os.path.join(self.dirname, filename)
+                        files = os.listdir(path)
+                        if files:
+                            path = os.path.join(path, files[0])
+                        else:
+                            # If the directory is empty skip it.
+                            # They are deleted elsewhere.
+                            continue
+                    age = time_now - last_access_time(path)
+
+                    # In normal case, the processus that created this
+                    # directory will delete it. However, if this processus
+                    # crashes, it will not be cleaned up.
+                    # As we don't know if this directory is still used,
+                    # we wait one week and suppose that the processus
+                    # crashed, and we take care of the clean-up.
+                    if age > min_age:
+                        to_del.append(os.path.join(self.dirname, filename))
+
+        if not to_del:
+            return
+        with compilelock.lock_ctx():
+            for f in to_del:
+                _rmtree(f,
+                        msg='old unversioned', level=logging.INFO,
+                        ignore_nocleanup=True)
 
     def _on_atexit(self):
         # Note: no need to call refresh() since it is called by clear_old().
-        with compilelock.lock_ctx():
-            self.clear_old()
-            self.clear_unversioned()
+
+        # Note: no need to take the lock. For unversioned files, we
+        # don't need it as they aren't shared. For old unversioned
+        # files, this happen rarely, so we take the lock only when
+        # this happen.
+
+        # Note: for clear_old(), as this happen unfrequently, we only
+        # take the lock when it happen.
```

Comment:
I think the above information is important enough to be included in the documentation of the methods clear_old() and clear_unversioned().